ide/rust/SyncRunner.kt

```diff
@@ -0,0 +1,115 @@
+class SyncRunner : Callable<Unit> {
+
+    @CommandLine.Option(names = ["--verbose", "-v"], required = false)
+    var verbose: Boolean = false
+
+    lateinit var logger: Logger
+    private lateinit var shell: Shell
+    private lateinit var buildWorkspaceDir: Path
+    lateinit var bazelOutputBase: Path
+
+    override fun call() {
+        logger = Logger(logLevel = if (verbose) DEBUG else ERROR)
+        shell = Shell(logger, verbose)
+        buildWorkspaceDir = Path(System.getenv("BUILD_WORKSPACE_DIRECTORY"))
+        bazelOutputBase = Path(shell.execute(listOf(BAZEL, INFO, "output_base"), buildWorkspaceDir).outputString().trim())
+        val rustTargets = rustTargets(buildWorkspaceDir)
+        validateBuildWorkspace(rustTargets)
+        loadRustToolchainAndExternalDeps(rustTargets)
+        rustRepoPaths().forEach { sync(it) }
+    }
+
+    private fun validateBuildWorkspace(rustTargets: List<String>) {
+        shell.execute(command = listOf(BAZEL, BUILD) + rustTargets + "--build=false", baseDir = buildWorkspaceDir)
+    }
+
+    private fun loadRustToolchainAndExternalDeps(rustTargets: List<String>) {
+        shell.execute(command = listOf(BAZEL, BUILD) + rustTargets, baseDir = buildWorkspaceDir, throwOnError = false)
+    }
+
+    private fun rustTargets(repoPath: Path): List<String> {
+        return shell.execute(listOf(BAZEL, QUERY, "kind(rust_*, //...)"), repoPath)
+            .outputString().split("\n").filter { it.isNotBlank() }
+    }
+
+    private fun rustRepoPaths(): List<Path> {
+        // e.g: [/Users/root/workspace/typedb-client-rust, /private/var/_bazel_root_/123abc/external/vaticle_typedb_protocol]
+        return listOf(buildWorkspaceDir) + shell.execute(listOf(BAZEL, QUERY, "kind(rust_*, deps(//...))", "--output=package"), buildWorkspaceDir)
+            .outputString().split("\n")
+            .filter { it.startsWith("@vaticle_") }
+            .map { it.split("@")[1].split("//")[0] }
+            .map { bazelOutputBase.resolve("external").resolve(it) }
+    }
+
+    private fun sync(repoPath: Path) {
+        logger.debug { "Syncing $repoPath" }
+        cleanupOldSyncInfo(repoPath)
+        runSyncInfoAspect(repoPath)
+        CargoManifestGenerator(repoPath.toFile(), shell).generateManifests()
+        logger.debug { "Sync completed in $repoPath" }
+    }
+
+    private fun cleanupOldSyncInfo(repoPath: Path) {
+        logger.debug { "Cleaning up old sync info under $repoPath" }
+        val bazelBin = File(shell.execute(listOf(BAZEL, INFO, "bazel-bin"), repoPath).outputString().trim())
+        bazelBin.listFilesRecursively().filter { it.name.endsWith(".ide-sync.properties") }.forEach { it.delete() }
+    }
+
+    private fun runSyncInfoAspect(repoPath: Path) {
+        val rustTargets = rustTargets(repoPath)
+        shell.execute(listOf(BAZEL, BUILD) + rustTargets + listOf(
+            "--aspects",
+            "@vaticle_dependencies//ide/rust:sync_aspect.bzl%rust_ide_sync_aspect",
```

Comment:
Again, these may break when we refactor the repo. Let's see if we can test them or gather them into a single "constant" or "config" object at least.